Improve docs on functions

danutsu opened the pull request:
While learning slint, I noticed that a lot of behaviors (in this case about functions) aren't clearly stated in the documentation. So I'm trying to document what I've learned and hopefully make it easier for the next person.

**I've tested this mostly by experiment**, meaning I'm not sure what's Working as Intended / by-design and what's Working as Implemented / by-chance. Actually that's a problem I'm hoping to address by documenting things more explicitly, but I need the core developers to tell me if something I've documented is not actually supposed to work that way.

Also, I don't know if you have specific standards/style regarding documentation that I should be following. If you do, let me know, I'm happy to read/apply them :)

There are more things I'd like to see in this file, in particular:

- typical use cases for functions
- more info/examples on how to use protected functions
- differences between functions and callbacks and guidance on when to use each

But I don't feel I know enough about these topics myself yet, so I'm starting with this.

ogoffart commented on docs/reference/src/language/syntax/functions.md:
```diff
@@ -14,27 +32,122 @@ Functions can be annotated with the `pure` keyword.
+
+The two types of calls function mostly the same. The only differences are:
+
+- Functions can be called without a target only from the element that defines them and its children.
+That means a function defined at the root of a component can be called from anywhere within that component
+while a function defined in an element can be called from that element and any child elements of it.
+- When a function is overridden, calling it with a target allows the caller to choose the exact version
+it wants to call (see Function Overriding below).
```
In fact, this is the same scoping as properties and callback.

danutsu replied:
Done: I mentioned this explicitly now.

ogoffart commented on docs/reference/src/language/syntax/functions.md:
```diff
@@ -14,27 +32,122 @@ Functions can be annotated with the `pure` keyword.
+export component CallsFunction {
+    property <int> test: my-friend.double(1);
+
+    my-friend: HasFunction {
```
```suggestion
    my-friend := HasFunction {
```

danutsu replied:
Thanks, fixed!

ogoffart commented on docs/reference/src/language/syntax/functions.md:
```diff
@@ -14,27 +32,122 @@ Functions can be annotated with the `pure` keyword.
+Functions, even marked public, cannot be exported and cannot be called from backend code (in Rust, C++,
+JS, etc.). Declare a [callback](callbacks.md) which can call the function.
+
+## Function overriding
```
I'm not sure this is technically overriding. It is just name scoping.
Same is true for callback or properties with the same name.

danutsu replied:
It's not really "scoping" either, I think, because the scope of the function is still the whole component. But I reorganized things a bit and describe this as name resolution rules, and instead of overriding I talk about "shadowing" -- this is similar to variable shadowing in Rust (https://doc.rust-lang.org/rust-by-example/variable_bindings/scope.html).

I think it's clearer this way, please take a look!

ogoffart reviewed (commented):
Thanks a lot. This is definitively an improvement!

This introduces the notion of "target" which is kind of new in the documentation. I think "element name" might be a better term.

ogoffart commented:
The CI is failling in the `doctests` test which checks that the slint snippets in the documentation are valid. You need to mark them as ignored, or they need to import the right types

```
failures:

---- docsⳇreferenceⳇsrcⳇlanguageⳇsyntaxⳇfunctionsᐧmd::line_125 stdout ----
thread 'docsⳇreferenceⳇsrcⳇlanguageⳇsyntaxⳇfunctionsᐧmd::line_125' panicked at /home/runner/work/slint/slint/target/debug/build/doctests-1e236a2bc24032cd/out/test_functions.rs:480:568:
called `Result::unwrap()` on an `Err` value: "Unknown type VerticalBox"
stack backtrace:
   0: rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::result::unwrap_failed
   3: doctests::docsⳇreferenceⳇsrcⳇlanguageⳇsyntaxⳇfunctionsᐧmd::line_125
   4: doctests::docsⳇreferenceⳇsrcⳇlanguageⳇsyntaxⳇfunctionsᐧmd::line_125::{{closure}}
   5: core::ops::function::FnOnce::call_once
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

---- docsⳇreferenceⳇsrcⳇlanguageⳇsyntaxⳇfunctionsᐧmd::line_16 stdout ----
thread 'docsⳇreferenceⳇsrcⳇlanguageⳇsyntaxⳇfunctionsᐧmd::line_16' panicked at /home/runner/work/slint/slint/target/debug/build/doctests-1e236a2bc24032cd/out/test_functions.rs:456:215:
called `Result::unwrap()` on an `Err` value: "Parse error: expected a top-level item such as a component, a struct, or a global"
stack backtrace:
   0: rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::result::unwrap_failed
   3: doctests::docsⳇreferenceⳇsrcⳇlanguageⳇsyntaxⳇfunctionsᐧmd::line_16
   4: doctests::docsⳇreferenceⳇsrcⳇlanguageⳇsyntaxⳇfunctionsᐧmd::line_16::{{closure}}
   5: core::ops::function::FnOnce::call_once
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

---- docsⳇreferenceⳇsrcⳇlanguageⳇsyntaxⳇfunctionsᐧmd::line_78 stdout ----
thread 'docsⳇreferenceⳇsrcⳇlanguageⳇsyntaxⳇfunctionsᐧmd::line_78' panicked at /home/runner/work/slint/slint/target/debug/build/doctests-1e236a2bc24032cd/out/test_functions.rs:468:347:
called `Result::unwrap()` on an `Err` value: "Syntax error: expected ';'"
stack backtrace:
   0: rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::result::unwrap_failed
   3: doctests::docsⳇreferenceⳇsrcⳇlanguageⳇsyntaxⳇfunctionsᐧmd::line_78
   4: doctests::docsⳇreferenceⳇsrcⳇlanguageⳇsyntaxⳇfunctionsᐧmd::line_78::{{closure}}
   5: core::ops::function::FnOnce::call_once
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

---- docsⳇreferenceⳇsrcⳇlanguageⳇsyntaxⳇfunctionsᐧmd::line_97 stdout ----
thread 'docsⳇreferenceⳇsrcⳇlanguageⳇsyntaxⳇfunctionsᐧmd::line_97' panicked at /home/runner/work/slint/slint/target/debug/build/doctests-1e236a2bc24032cd/out/test_functions.rs:474:412:
called `Result::unwrap()` on an `Err` value: "Syntax error: expected ';'"
stack backtrace:
   0: rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::result::unwrap_failed
   3: doctests::docsⳇreferenceⳇsrcⳇlanguageⳇsyntaxⳇfunctionsᐧmd::line_97
   4: doctests::docsⳇreferenceⳇsrcⳇlanguageⳇsyntaxⳇfunctionsᐧmd::line_97::{{closure}}
   5: core::ops::function::FnOnce::call_once
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

---- docsⳇreferenceⳇsrcⳇlanguageⳇsyntaxⳇfunctionsᐧmd::line_40 stdout ----
thread 'docsⳇreferenceⳇsrcⳇlanguageⳇsyntaxⳇfunctionsᐧmd::line_40' panicked at /home/runner/work/slint/slint/target/debug/build/doctests-1e236a2bc24032cd/out/test_functions.rs:462:611:
called `Result::unwrap()` on an `Err` value: "Unknown type Button"
stack backtrace:
   0: rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::result::unwrap_failed
   3: doctests::docsⳇreferenceⳇsrcⳇlanguageⳇsyntaxⳇfunctionsᐧmd::line_40
   4: doctests::docsⳇreferenceⳇsrcⳇlanguageⳇsyntaxⳇfunctionsᐧmd::line_40::{{closure}}
   5: core::ops::function::FnOnce::call_once
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.


failures:
    docsⳇreferenceⳇsrcⳇlanguageⳇsyntaxⳇfunctionsᐧmd::line_125
    docsⳇreferenceⳇsrcⳇlanguageⳇsyntaxⳇfunctionsᐧmd::line_16
    docsⳇreferenceⳇsrcⳇlanguageⳇsyntaxⳇfunctionsᐧmd::line_40
    docsⳇreferenceⳇsrcⳇlanguageⳇsyntaxⳇfunctionsᐧmd::line_78
    docsⳇreferenceⳇsrcⳇlanguageⳇsyntaxⳇfunctionsᐧmd::line_97
```

danutsu commented:
Thanks! I fixed the tests (hadn't noticed them) and I reorganized things a bit to make things clearer. I also added a section on name resolution inside the function code, because I was a bit unsure about it myself -- I could see it being implemented both ways!

Please take another look :)

ogoffart commented on docs/reference/src/language/syntax/functions.md:
```diff
@@ -14,27 +35,158 @@ Functions can be annotated with the `pure` keyword.
 - A function annotated with `protected` can only be accessed by components that directly inherit from it.
 
-## Example
+Functions, even marked public, cannot be exported and cannot be called from backend code (in Rust, C++,
+JS, etc.). Declare a [callback](callbacks.md) which can call the function.
```
Function can be called from backend code. (invoke_...)
The difference with callback is that they cannot be replaced with a callback handler.

danutsu replied:
Also, your point about the difference between this and callbacks made me take a stab at a "functions vs callbacks" section, could you look over it and let me know if I got anything wrong?

ogoffart replied:
Indeed, falling function from native code is missing from that documentation 🙈

ogoffart commented on docs/reference/src/language/syntax/functions.md:
```diff
@@ -14,27 +35,158 @@ Functions can be annotated with the `pure` keyword.
+Functions, even marked public, cannot be exported and cannot be called from backend code (in Rust, C++,
+JS, etc.). Declare a [callback](callbacks.md) which can call the function.
+
+## Name resolution in function code
```
Name resolution in function code is actually the same as in callback handler and property bindings (defined in https://github.com/slint-ui/slint/issues/273 and arguably not well documented)
 1. function or callback arguments
 2. id of elements (including `self`, `root`, and `parent`)
 3. property or model/index name in scope.
 4. imported enums or globals
 5. builtin namespaces (Math, Colors, Key)
 6. Return specific lookup depending on the return type of a function / callback, or the type of a property binding. If this is an enum, then enum names are in scope, if this is a brush or color, then the Colors namespace is in scope.
 7. Builtin functions (from Math or Colors namespace, or `debug` or `animation-tick`)

danutsu replied:
Ah, OK. I removed the second section. I think it would make sense to document this somewhere in the .slint docs to make it more discoverable -- of course it doesn't make sense to add it to the functions doc -- I'm thinking somewhere more specific, like a separate page under syntax? (then functions, callbacks, properties, etc. could all link to it). The github issue is very hard for anyone to discover :)

I can do that if you want (but also in a separate PR, keeping this one focused)

ogoffart reviewed (commented):
Thanks for the edits. Added a few comments.

danutsu commented:
Thank you, please take another look :smile:

danutsu commented on docs/reference/src/language/syntax/functions.md:
```diff
@@ -14,27 +35,152 @@ Functions can be annotated with the `pure` keyword.
+- Callback visiblity is always similar to `public` functions
+
+In general, the biggest reason to use callbacks is to be able to handle them from the backend code. If
+that is not needed, using a callback or a function is a matter of personal choice.
```
I kind of thought so myself, but didn't feel comfortable saying this without input from the core team. Updated to say "use a function" :)

tronical commented on docs/reference/src/language/syntax/functions.md:
```diff
@@ -14,27 +35,152 @@ Functions can be annotated with the `pure` keyword.
+There are a lot of similarities between functions and [callbacks](callbacks.md):
+
+- They are both callable blocks of logic/code
+- They can be invoked similarly
```
Aren't they invoked not only similarly, but in exactly the same way?

danutsu replied:
Done -- see, as someone just learning Slint, I'm a little bit afraid of making calls like that. As far as I know, they are, but I'm not an expert, so this was just me hedging :stuck_out_tongue:

tronical commented on docs/reference/src/language/syntax/functions.md:
```diff
@@ -14,27 +35,152 @@ Functions can be annotated with the `pure` keyword.
+- They can both have parameters and return values
+- They can both be declared `pure`
+
+But there are also key differences:
```
My immediate question here would be: If these are the key differences, where can I read about the remaining differences? I suggest to leave that out for now :)

```suggestion
But there are also differences:
```

danutsu replied:
Done! (to be fair, "what else is different" is also a question I have and I honestly don't know the answer to it)

tronical commented on docs/reference/src/language/syntax/functions.md:
```diff
@@ -14,27 +35,152 @@ Functions can be annotated with the `pure` keyword.
+- Callbacks have a special syntax for declaring aliases using the two-way binding operator `<=>`
+- Callback visiblity is always similar to `public` functions
+
+In general, the biggest reason to use callbacks is to be able to handle them from the backend code. If
+that is not needed, we recommend using a function.
```
Minor simplification:

```suggestion
In general, the biggest reason to use callbacks is to be able to handle them from the backend code. Use
a function if that is not needed.
```

danutsu replied:
Done.

tronical reviewed (approved):
A couple of suggestions inline. Overall I think this is an improvement, but I wouldn't have put so much emphasis on the lookup.

danutsu commented on docs/reference/src/language/syntax/functions.md:
```diff
@@ -1,10 +1,31 @@
 
-Functions in Slint are declared using the `function` keyword.
+Similarly to other programming languages, functions in Slint are way to name, organize and reuse
+a piece of logic/code.
```
Done.

danutsu commented on docs/reference/src/language/syntax/functions.md:
```diff
@@ -1,10 +1,31 @@
-Each function can have parameters which are declared within parentheses, following the format `name : type`.
-These parameters can be referenced by their names within the function body.
+Functions can be defined as part of a component, or as part of an element within a component. Functions
+are always part of a component: it is not possible to declare global (top-level) functions, or to
```
Done, good point.

danutsu commented on docs/reference/src/language/syntax/functions.md:
```diff
@@ -1,10 +1,31 @@
+declare them as part of a struct or enum. It is also not possible to nest functions within other
+functions.
+
+## Declaring a function
```
Done.
